core/src/test/scala/dagr/core/exec/SimpleTaskCacheTest.scala

```diff
@@ -0,0 +1,473 @@
+    val executor1 = Executor(experimentalExecution=true, resources=SystemResources.infinite)
+    val pipeline1 = new SleepyPipeline(failTaskIndex=failTaskIndex, shouldFail=true)
+    executor1.withReporter(new ExecutionLogger(replayLog))
+    if (failTaskIndex.nonEmpty) executor1.execute(pipeline1) should be > 0
+    else executor1.execute(pipeline1) shouldBe 0
+
+    // Execute it a second time, with the replay log.  If `failTaskIndex` has any values, those tasks will not fail.
+    val executor2 = Executor(experimentalExecution=true, resources=SystemResources.infinite)
+    val pipeline2 = new SleepyPipeline(failTaskIndex=Set.empty, shouldFail=false)
+    executor2.withReporter(TaskCache(replayLog))
+    executor2.execute(pipeline2) shouldBe 0
+    Logger.out = tmpOut
+  }
+
+  "SimpleTaskCache" should "replay a complicated workflow that succeeded all tasks" in {
+    executeAndReplay(failTaskIndex=Set.empty)
+  }
+
+  it should "replay a complicated workflow whose first task failed" in {
+    executeAndReplay(failTaskIndex=Set(0))
+  }
+
+  it should "replay a complicated workflow who has a single failing task somewhere in the middle" in {
+    executeAndReplay(failTaskIndex=Set(50))
+  }
+
+  it should "replay a complicated workflow whose last task failed" in {
+    executeAndReplay(failTaskIndex=Set(99))
+  }
+
+  it should "replay a complicated workflow where all tasks fail" in {
+    executeAndReplay(failTaskIndex=Seq.range(0, 100).toSet)
+  }
+
+  it should "replay a complicated workflow where tasks fail" in {
+    val tmpOut = Logger.out
+    Logger.out = new PrintStream(new OutputStream {
+      override def write(b: Int) = Unit
+    })
+
+    var reportLog = {
+      val path = Files.createTempFile("report.", ".log")
+      path.toFile.deleteOnExit()
+      path
+    }
+
+    var replayLog = {
+      val path = Files.createTempFile("replay.", ".log")
+      path.toFile.deleteOnExit()
+      path
+    }
+
+    // fail every task once!
+    Seq.range(0, 100).foreach { failTaskIndex =>
+      val executor = Executor(experimentalExecution=true, resources=SystemResources.infinite)
+      val pipeline = new SleepyPipeline(failTaskIndex=Set(failTaskIndex), shouldFail=true)
+      executor.withReporter(new ExecutionLogger(reportLog))
+      if (0 < failTaskIndex) executor.withReporter(TaskCache(replayLog))
+      executor.execute(pipeline) should be > 0
+      // swap the replay and report log
+      val tmp   = replayLog
+      replayLog = reportLog
+      reportLog = tmp
+    }
+
+    // run it one last time, without failing
+    {
+      val executor = Executor(experimentalExecution=true, resources=SystemResources.infinite)
+      val pipeline = new SleepyPipeline(failTaskIndex=Set.empty, shouldFail=false)
+      executor.withReporter(TaskCache(replayLog))
+      executor.execute(pipeline) shouldBe 0
+    }
+    Logger.out = tmpOut
+  }
+}
```

Comment:
TODO: add a test of a pipeline within a pipeline within a pipeline, such that the middle-level pipeline produces different pipelines on the second execution, and so the middle and bottom level pipelines are forced to execute.